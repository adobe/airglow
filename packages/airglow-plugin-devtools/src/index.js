/*
Copyright 2019 Adobe. All rights reserved.
This file is licensed to you under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License. You may obtain a copy
of the License at http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software distributed under
the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
OF ANY KIND, either express or implied. See the License for the specific language
governing permissions and limitations under the License.
*/

import React from 'react';
import { POST_ENHANCER, COMPONENT, COMPOSER } from 'airglow';
import Devtools from './devtools';

export default function devtoolsPlugin({
  reduxDevtools = false,
  reduxBrowser = false
} = {}) {
  return (engine) => {
    if (reduxDevtools) {
      engine.plugin(POST_ENHANCER, Devtools.instrument());
      engine.plugin(COMPONENT, <Devtools key="c" />);
    }
    if (reduxBrowser) {
      engine.plugin(COMPOSER, browserCompose);
    }
  };
}

/* eslint-disable no-underscore-dangle */
const browserHook = () => global.window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__;

const browserCompose = compose =>
  (browserHook() ? browserHook()({}) : compose);
