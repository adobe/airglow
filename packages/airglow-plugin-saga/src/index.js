/*
Copyright 2019 Adobe. All rights reserved.
This file is licensed to you under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License. You may obtain a copy
of the License at http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software distributed under
the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
OF ANY KIND, either express or implied. See the License for the specific language
governing permissions and limitations under the License.
*/

import createSagaMiddleware from 'redux-saga';
import { FEATURE, MIDDLEWARE, BOOTSTRAP_MODULE } from 'airglow';

export const SAGA = 'wire_sagas';

export default function sagaPlugin() {
  const middleware = createSagaMiddleware();

  return (engine) => {
    engine.plugin(MIDDLEWARE, middleware);
    engine.plugin(FEATURE, sagaFeature(middleware));
    engine.plugin(BOOTSTRAP_MODULE, bootstrapSagas(middleware));
  };
}

const sagaFeature = middleware => lookup =>
  lookup(SAGA).forEach(
    plugin => middleware.run(plugin)
  );

const bootstrapSagas = middleware => ({ sagas }) => {
  if (!sagas) { return; }
  const runSagas = Array.isArray(sagas) ? sagas : [sagas];
  runSagas.forEach(saga => middleware.run(saga));
};
