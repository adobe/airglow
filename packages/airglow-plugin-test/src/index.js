/*
Copyright 2019 Adobe. All rights reserved.
This file is licensed to you under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License. You may obtain a copy
of the License at http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software distributed under
the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
OF ANY KIND, either express or implied. See the License for the specific language
governing permissions and limitations under the License.
*/

import { REDUCER } from 'airglow';
import { slice } from '@airglow/reducers';
import { last } from 'ramda';
import reducer from './reducer';

const STORE_KEY = '_test_dispatches';

export default function testPlugin() {
  return (engine) => {
    engine.plugin(REDUCER, slice(STORE_KEY).with(reducer));
  };
}

export const lastDispatch = state => last(state[STORE_KEY]);
export const dispatchList = state => state[STORE_KEY];
