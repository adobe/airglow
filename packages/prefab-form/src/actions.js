/*
Copyright 2019 Adobe. All rights reserved.
This file is licensed to you under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License. You may obtain a copy
of the License at http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software distributed under
the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
OF ANY KIND, either express or implied. See the License for the specific language
governing permissions and limitations under the License.
*/

import { createAction } from '@airglow/reducers';

export const VALUE_CHANGE      = 'PREFAB_FORM_FIELD_VALUE';
export const BLUR              = 'PREFAB_FORM_FIELD_BLUR';
export const FOCUS             = 'PREFAB_FORM_FIELD_FOCUS';
export const RESET             = 'PREFAB_FORM_FIELD_RESET';
export const FIELD_SUBMIT      = 'PREFAB_FORM_FIELD_SUBMIT';
export const SUBMIT            = 'PREFAB_FORM_SUBMIT';

const mapConstruct = (construct, value) => ({ construct, value });

export const change = createAction(VALUE_CHANGE, mapConstruct);
export const focus = createAction(FOCUS, mapConstruct);
export const blur = createAction(BLUR, mapConstruct);
export const reset = createAction(RESET, mapConstruct);
export const submitField = createAction(FIELD_SUBMIT, mapConstruct);
export const submit = createAction(SUBMIT, mapConstruct);
