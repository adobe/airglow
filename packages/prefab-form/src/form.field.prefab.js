/*
Copyright 2019 Adobe. All rights reserved.
This file is licensed to you under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License. You may obtain a copy
of the License at http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software distributed under
the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
OF ANY KIND, either express or implied. See the License for the specific language
governing permissions and limitations under the License.
*/

import * as R from 'ramda';
import reduce from './reducers';
import formFieldSelector from './selectors/form.field.selectors';

export default ({
  name,
  fieldName,
  externalName,
  defaultValue,
  fieldType = 'text',
  ...additional
}) => {
  const selectorInternal = formFieldSelector(name);
  selectorInternal.name = name;
  selectorInternal.externalName = externalName || name;
  selectorInternal.fieldName = fieldName || name;
  selectorInternal.defaultValue = defaultValue;
  selectorInternal.fieldType = fieldType;
  selectorInternal.reduce = reduce;

  R.forEachObjIndexed((v, k) => { selectorInternal[k] = v; }, additional);

  return { [name]: selectorInternal };
};
