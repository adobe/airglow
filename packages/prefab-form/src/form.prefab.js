/*
Copyright 2019 Adobe. All rights reserved.
This file is licensed to you under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License. You may obtain a copy
of the License at http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software distributed under
the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
OF ANY KIND, either express or implied. See the License for the specific language
governing permissions and limitations under the License.
*/

import * as R from 'ramda';
import formField from './form.field.prefab';
import formSelector from './selectors/form.selectors';

export default ({
  autocorrect = 'enabled',
  fields,
  name,
  sourceSelector,
  whenToValidate = 'submit',
  ...additional
}) => {
  const { prefabs, references } = processFields({
    autocorrect,
    fields,
    name,
    sourceSelector,
    whenToValidate
  });

  const selectorInternal = formSelector({ fields: references, name, ...additional });

  selectorInternal.whenToValidate = whenToValidate;
  selectorInternal.autocorrect = autocorrect;
  selectorInternal.formName = name;
  selectorInternal.type = 'form';

  R.forEachObjIndexed(
    (v, k) => { selectorInternal[k] = selectorInternal[k] || v; },
    additional
  );

  return { [name]: selectorInternal, ...prefabs };
};

const processFields = ({ fields, name: parentName, sourceSelector: parentSelector, ...config }) => {
  let prefabs = {};
  let references = {};
  R.forEachObjIndexed(
    ({ type, sourceSelector, fieldName, ...options }, keyName) => {
      const customSource = parentSelector ? state => parentSelector(state)[keyName] : null;
      const storeName = `${parentName}#${keyName}`;
      const useFieldName = fieldName || `${keyName}Field`;
      const results = formField({
        name: storeName,
        externalName: keyName,
        fieldName: useFieldName,
        sourceSelector: sourceSelector || customSource,
        ...config,
        ...options
      });
      prefabs = R.assoc(storeName, results[storeName], prefabs);
      references = R.assoc(useFieldName, results[storeName], references);
    },
    fields
  );
  return {
    prefabs,
    references
  };
};
