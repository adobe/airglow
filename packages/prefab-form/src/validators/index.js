/*
Copyright 2019 Adobe. All rights reserved.
This file is licensed to you under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License. You may obtain a copy
of the License at http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software distributed under
the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
OF ANY KIND, either express or implied. See the License for the specific language
governing permissions and limitations under the License.
*/

import isValidJson from './is.valid.json';

const makeValidator = (check, error) => value => (
  check(value) ? false : error
);

const emailRegex = /^(([^<>()[\]\\.,;:\s@"]+(\.[^<>()[\]\\.,;:\s@"]+)*)|(".+"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/;
const ipRegex = /^(?!0)(?!.*\.$)((1?\d?\d|25[0-5]|2[0-4]\d)(\.|$)){4}$/;
const uriRegex = new RegExp('^(https?:\\/\\/)?' // protocol
  + '((([a-z\\d]([a-z\\d-]*[a-z\\d])*)\\.)+[a-z]{2,}|' // domain name
  + '((\\d{1,3}\\.){3}\\d{1,3}))' // OR ip (v4) address
  + '(\\:\\d+)?(\\/[-a-z\\d%_.~+]*)*' // port and path
  + '(\\?[;&a-z\\d%_.~+=-]*)?' // query string
  + '(\\#[-a-z\\d_]*)?$', 'i'); // fragment locator

const isLength = length => value => !value || value.length === length;
const isMaxLength = max => value => !value || value.length <= max;
const isMinLength = min => value => !value || value.length >= min;
const matches = regex => value => !value || value.match && value.match(regex);
const isAlphaNum = matches(/^[a-zA-Z0-9_]*$/);
const isEmail = matches(emailRegex);
const isIP = matches(ipRegex);
const isUri = matches(uriRegex);

const isMinValue = min => value => value >= min;
const isMaxValue = max => value => value <= max;
const isGreater = min => value => value > min;
const isLess = max => value => value < max;
const isNumber = v => typeof v === 'number';
const isString = v => !v || typeof v === 'string';
const hasPrecision = p => v => `${v}` === v.toFixed(p);

export const json = makeValidator(isValidJson, 'error.string.json');
export const minLength = min => makeValidator(isMinLength(min), 'error.string.min');
export const maxLength = max => makeValidator(isMaxLength(max), 'error.string.max');
export const length = l => makeValidator(isLength(l), 'error.string.length');
export const regex = rx => makeValidator(matches(rx), 'error.string.regex');
export const alphanum = makeValidator(isAlphaNum, 'error.string.alphanum');
export const email = makeValidator(isEmail, 'error.string.email');
export const ip = makeValidator(isIP, 'error.string.ip');
export const uri = makeValidator(isUri, 'error.string.uri');
export const string = makeValidator(isString, 'error.string.base');

export const min = m => makeValidator(isMinValue(m), 'error.number.min');
export const max = m => makeValidator(isMaxValue(m), 'error.number.max');
export const greater = m => makeValidator(isGreater(m), 'error.number.greater');
export const less = m => makeValidator(isLess(m), 'error.number.less');
export const integer = makeValidator(Number.isInteger, 'error.number.integer');
export const precision = p => makeValidator(hasPrecision(p), 'error.number.precision');
export const number = makeValidator(isNumber, 'error.number.base');
