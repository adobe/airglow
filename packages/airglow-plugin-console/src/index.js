/*
Copyright 2019 Adobe. All rights reserved.
This file is licensed to you under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License. You may obtain a copy
of the License at http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software distributed under
the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
OF ANY KIND, either express or implied. See the License for the specific language
governing permissions and limitations under the License.
*/

import * as R from 'ramda';
import { REDUCER, BOOTSTRAP_MODULE, BOOTSTRAP_PLUGIN } from 'airglow';
import { slice, createAction } from '@airglow/reducers';
import { BOOTSTRAP_CONSOLE } from './actions';
import reducer from './reducers';

const STORE_KEY = 'console-commands';
const doBootstrap  = createAction(BOOTSTRAP_CONSOLE);

export default function prefabPlugin(config) {
  return (engine) => {
    engine.plugin(REDUCER, slice(STORE_KEY).with(reducer));
    engine.plugin(BOOTSTRAP_MODULE, bootstrap(engine));
    engine.plugin(BOOTSTRAP_PLUGIN, bootstrapPlugin(engine, config));
  };
}

const logHelp = ({ state }) => {
  R.forEachObjIndexed(
    ({ description }, key) => {
      console.info(`${key}(): ${description}`); // eslint-disable-line no-console
    },
    state['console-commands']
  );
};

const bootstrapPlugin = (
  engine, { help = 'help', helpDescription = 'Display this help menu', consoleCommands }
) => ({ dispatch }) => {
  bootstrap(engine)({
    dispatch,
    consoleCommands: {
      [help]: { description: helpDescription, command: logHelp },
      ...consoleCommands
    }
  });
};

const bootstrap = engine => ({ dispatch, consoleCommands }) => {
  if (consoleCommands) {
    dispatch(doBootstrap(consoleCommands));
    R.forEachObjIndexed(
      ({ command }, key) => {
        window[key] = () => command({ state: engine.getStore().getState(), dispatch });
      },
      consoleCommands
    );
  }
};
