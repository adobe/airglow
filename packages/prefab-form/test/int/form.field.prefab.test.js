/*
Copyright 2019 Adobe. All rights reserved.
This file is licensed to you under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License. You may obtain a copy
of the License at http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software distributed under
the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
OF ANY KIND, either express or implied. See the License for the specific language
governing permissions and limitations under the License.
*/

import { createStore } from 'redux';
import { slice, call } from '@airglow/reducers';
import prefab, { reducer, BOOTSTRAP_PREFAB } from '@airglow/prefab';
import '../../src';

let store;
let prefabs;

describe('FormFieldPrefabIntegration', () => {
  beforeEach(() => {
    store = createStore(
      call(
        slice('prefab').with(reducer)
      )
    );

    prefabs = prefab({
      pizza: {
        type: 'formField',
        defaultValue: 'cheese',
        validator: v => (v.indexOf('z') !== -1 ? 'error.z' : false),
        whenToValidate: 'blur'
      },
      size: {
        type: 'formField',
        fieldType: 'number',
        defaultValue: 2,
        max: 15,
        autocorrect: 'immediate'
      }
    });
    store.dispatch({
      type: BOOTSTRAP_PREFAB,
      payload: prefabs
    });
  });

  it('should correctly select the initial values', function () {
    expect(prefabs.pizza.value(store.getState())).toBe('cheese');
    expect(prefabs.size.value(store.getState())).toBe(2);
  });

  it('should correctly update the value', function () {
    const handlers = prefabs.pizza.handlers(store.dispatch);
    handlers.onChange('meat');
    expect(prefabs.pizza.value(store.getState())).toBe('meat');
  });

  it('should autocorrect', () => {
    const handlers = prefabs.size.handlers(store.dispatch);
    handlers.onChange(52);
    expect(prefabs.size.value(store.getState())).toBe(15);
  });

  it('should validate', () => {
    expect(prefabs.pizza.isInvalid(store.getState())).toBe(false);
    const handlers = prefabs.pizza.handlers(store.dispatch);
    handlers.onChange('meatz');
    expect(prefabs.pizza.state(store.getState())).toMatchSnapshot();

    handlers.onBlur();
    expect(prefabs.pizza.state(store.getState())).toMatchSnapshot();
    expect(prefabs.pizza.isInvalid(store.getState())).toBe('error.z');
  });

  it('can be dirty', () => {
    expect(prefabs.pizza.isDirty(store.getState())).toBe(false);
    const handlers = prefabs.pizza.handlers(store.dispatch);
    handlers.onChange('meat');
    expect(prefabs.pizza.isDirty(store.getState())).toBe(true);
  });

  it('resets', () => {
    const handlers = prefabs.pizza.handlers(store.dispatch);
    handlers.onChange('meat');
    handlers.onReset();
    expect(prefabs.pizza.value(store.getState())).toBe('cheese');
  });
});
