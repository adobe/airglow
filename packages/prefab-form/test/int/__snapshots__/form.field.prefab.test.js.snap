// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormFieldPrefabIntegration should validate 1`] = `
Object {
  "dirty": true,
  "error": false,
  "invalidMsg": "error.z",
  "required": false,
  "value": "meatz",
}
`;

exports[`FormFieldPrefabIntegration should validate 2`] = `
Object {
  "dirty": true,
  "error": "error.z",
  "invalidMsg": "error.z",
  "required": false,
  "value": "meatz",
}
`;
